### PR TITLE
Verify controller attached volume watch

### LINKED ISSUE #28141: Allow Kubelet Node Status to be updated on demand
The kubelet volume manager depends on the `VolumesInUse` field of the node status being updated before it can proceed with attaching a volume. Since node updates happen periodically on kubelet it can take as much as 10 seconds before the field is updated.

Consider a force sync for the node status update code, something like:

``` GO
    select {
     case <- forceSync
     case <-time.After(10*time.Second)
     }
    tryUpdateNodestatus()
```

Where the `forceSync` channel can be passed to the volume manager to allow it to force updates faster than 10 seconds when needed.

### PR DESCRIPTION
**Pull request by shengnuo**
**What type of PR is this?**
This PR reopens https://github.com/kubernetes/kubernetes/pull/45344, which was closed by Author due to Community's lack of interest. 

/kind design

**What this PR does / why we need it**:
*ORIGINAL PR DESCRIPTION*

```
This PR changes VerifyControllerAttachedVolume from the kubelet volume manager to use a watch instead of a single get for node status.

VerifyControllerAttachedVolume is an asynchronous operation which returned pretty fast with an error in and thus exponential backoff hit pretty hard here. This PR changes the operation to wait/watch for some time (1 minute) for the node status to report that the volume is attached and only returns an error after timeout.

It also moves the ReportedInUse check into the watch loop and thus makes us independent from the node status cache.

This PR is an alternative solution for the same problem I tried to fix in #40531. I would prefer the new solution.
```

This PR has been rebased on top of latest Master.

**Which issue(s) this PR fixes**:

Fixes #28141 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by shengnuo** [comment]
@codablock  FYI

**Comment by ehashman** [comment]
@kubernetes/sig-storage-pr-reviews FYI

**Comment by ehashman** [comment]
/hold

The issue this fixes is so old that I want to make sure we have the go-ahead from SIG Storage before spending too much time and energy on it.

**Comment by hsj1606** [comment]
any update on this?

File: pkg/volume/util/operationexecutor/operation_generator.go
Comment (jingxu97) [inline pkg/volume/util/operationexecutor/operation_generator.go:1333]: is that possible to share the same watch among all volumes for the same node?

**Comment by jingxu97** [comment]
assign @jingxu97 @chrishenzie

File: pkg/volume/util/operationexecutor/operation_generator.go
Comment (chrishenzie) [reply pkg/volume/util/operationexecutor/operation_generator.go:1333]: We could cache the watch object, but would all goroutines reading this channel be able to see all updates? Also we would need to consider how to handle timeouts across goroutines.

Comment (chrishenzie) [inline pkg/volume/util/operationexecutor/operation_generator.go:1333]: Does this watch return the initial object? If this function is called after volumes are attached and a watch is created, won't this operation timeout if no node changes were coming?

**Comment by tomfeigin** [comment]
Hey we also want this issue resolved, we have adjusted this PR for the newest kubernetes version and it saved us more than a few (and very expensive) seconds, we would be happy to help in anyway we can to advance this PR.

**Review by chrishenzie — CHANGES_REQUESTED** [review]
It looks like this PR requires a rebase, support for `bazel` was removed recently.

Something to keep in mind with this PR is the impact of creating many API watches. If we start many pods and volumes simultaneously and each kubelet creates API watches for each volume in use to verify, this could cause issues for heavily loaded API servers. Usually when the API server is under heavy load, it will start closing open watches which will result in an error and kubelet having to reschedule the operation.

Can you explore ways of potentially reducing the number of API watches we need to create in this case? Two examples that come to mind:

- Sharing the watch across goroutines instead of creating one per volume
  - We would likely need separate logic to manage the lifecycle of this API watch, it seems like this could cause issues if the API watch timeout is hit and recently added goroutines exit prematurely (watch closed) and then all of those operations require rescheduling
- Do we watch for the node object elsewhere in the kubelet codebase? Are there any relevant types (informers, listers, something else that manages API watches) for nodes that we could reuse here?

**Comment by chrishenzie** [comment]
This PR (https://github.com/kubernetes/kubernetes/pull/97834) may be of use to you. TL;DR:
- The CSI attacher was previously creating one API watch per `VolumeAttachment` to verify attachment succeeded, which would cause issues when there were many of them
- In clusters under heavy load with many volume attachments, the API server would prematurely close watches which would cause attachment logic to take much longer
- We tried ["dynamically hooking into" a SharedInformer for volume attachments](https://github.com/kubernetes/kubernetes/pull/97834#discussion_r553733640) which would allow clients to watch for changes on specific attachment IDs, but this solution introduced a lot of complexity, had race conditions, and was difficult to debug and maintain
- The workaround was to poll a shared cache (the volume attachment lister) using exponential backoff (see [`waitForVolumeAttachDetachStatusWithLister`](https://github.com/kubernetes/kubernetes/pull/97834/files#diff-227f84916ffb93ece42ccaec840af8ea265714440c15c45f42b08d6a427a57bfR467))

**Comment by MadhavJivrajani** [comment]
/remove-kind design
/kind feature

`kind/design` will soon be removed from `k/k` in favor of `kind/feature`. Relevant discussion can be found here: https://github.com/kubernetes/community/issues/5641

**Pull request closed without merging**
